lightningd/connect_control.c

```diff
@@ -522,6 +522,34 @@ static void handle_custommsg_in(struct lightningd *ld, const u8 *msg)
 	plugin_hook_call_custommsg(ld, NULL, p);
 }
 
+static void connectd_start_shutdown_reply(struct subd *connectd,
+					  const u8 *reply,
+					  const int *fds UNUSED,
+					  void *unused UNUSED)
+{
+	if (!fromwire_connectd_start_shutdown_reply(reply))
+		fatal("Bad connectd_start_shutdown_reply: %s",
+		      tal_hex(reply, reply));
+
+	/* Break out of loop now, so we can continue shutdown. */
+	log_debug(connectd->ld->log, "io_break: %s", __func__);
+	io_break(connectd);
+}
+
+void connectd_start_shutdown(struct subd *connectd)
+{
+	void *ret;
+	const u8 *msg = towire_connectd_start_shutdown(NULL);
+
+	subd_req(connectd, connectd, take(msg), -1, 0,
+		 connectd_start_shutdown_reply, NULL);
+
+	/* Wait for shutdown_reply */
+	ret = io_loop(NULL, NULL);
+	log_debug(connectd->ld->log, "io_loop: %s", __func__);
+	assert(ret == connectd);
```

Comment:
Does the assertion look like not always true in some corner cases? but the io loop to me is still a field fo cln that I'm deep understand, so not sure what the cause can be, looks like a not atomic state of the global `io_return_loop`?

```
Leaving base_dir /tmp/ltests-gxt9iwny intact, it still has test sub-directories with failure details: ['test_funding_cancel_race_1']
Leaving base_dir /tmp/ltests-os3lngj1 intact, it still has test sub-directories with failure details: ['test_hsmtool_secret_decryption_1']
--------------------------- Captured stderr teardown ---------------------------
lightningd: lightningd/connect_control.c:550: connectd_start_shutdown: Assertion `ret == connectd' failed.
lightningd: FATAL SIGNAL 6 (version 5a48536-modded)
Log dumped in crash.log.20230203064412
Lost connection to the RPC socket.Lost connection to the RPC socket.Lost connection to the RPC socket.Lost connection to the RPC socket.Lost connection to the RPC socket.Lost connection to the RPC socket.Lost connection to the RPC socket.Lost connection to the RPC socket.Lost connection to the RPC socket.Lost connection to the RPC socket.Lost connection to the RPC socket.Lost connection to the RPC socket.Lost connection to the RPC socket.
=========================== short test summary info ============================
ERROR tests/test_connection.py::test_funding_cancel_race - ValueError: 
============= 45 passed, 10 skipped, 1 error in 1847.54s (0:30:47) =============
```

Reply:
Yes.  Since we're marked shutdown, the json command core will keep io_breaking every time it drains the queue.  Which is fine: we just need to wait for connectd.